packages/hardhat-core/src/internal/artifacts.ts

```diff
@@ -372,6 +385,192 @@ export class Artifacts implements IArtifacts {
     return path.join(this._artifactsPath, sourceName, `${contractName}.json`);
   }
 
+  private async _getValidArtifactPathFromFullyQualifiedName(
+    fullyQualifiedName: string
+  ): Promise<string> {
+    const artifactPath =
+      this.formArtifactPathFromFullyQualifiedName(fullyQualifiedName);
+
+    const trueCaseArtifactPath = await this._trueCasePath(
+      path.relative(this._artifactsPath, artifactPath),
+      this._artifactsPath
+    );
+
+    if (trueCaseArtifactPath === null) {
+      return this._handleWrongArtifactForFullyQualifiedName(fullyQualifiedName);
+    }
+
+    if (artifactPath !== trueCaseArtifactPath) {
+      throw new HardhatError(ERRORS.ARTIFACTS.WRONG_CASING, {
+        correct: trueCaseArtifactPath,
+        incorrect: artifactPath,
+      });
+    }
+
+    return artifactPath;
+  }
+
+  private _getAllContractNamesFromFiles(files: string[]): string[] {
+    return files.map((file) => {
+      const fqn = this._getFullyQualifiedNameFromPath(file);
+      return parseFullyQualifiedName(fqn).contractName;
+    });
+  }
+
+  private _getAllFullyQualifiedNamesSync(): string[] {
+    const paths = this._getArtifactPathsSync();
+    return paths.map((p) => this._getFullyQualifiedNameFromPath(p)).sort();
+  }
+
+  private _handleWrongArtifactForFullyQualifiedName(
+    fullyQualifiedName: string
+  ): never {
+    const names = this._getAllFullyQualifiedNamesSync();
+
+    if (names.length === 0) {
+      throw new HardhatError(ERRORS.ARTIFACTS.NOT_FOUND, {
+        fullyQualifiedName,
+      });
+    }
+
+    const similarNames = this._getSimilarContractNames(
+      fullyQualifiedName,
+      names
+    );
+
+    throw new HardhatError(ERRORS.ARTIFACTS.TYPO_SUGGESTION, {
+      contractName: fullyQualifiedName,
+      similarNames: similarNames.join(os.EOL),
+    });
+  }
+
+  private _handleWrongArtifactForContractName(
+    contractName: string,
+    files: string[]
+  ): never {
+    const names = this._getAllContractNamesFromFiles(files);
+
+    if (names.length === 0) {
+      throw new HardhatError(ERRORS.ARTIFACTS.NOT_FOUND, {
+        contractName,
+      });
+    }
+
+    const similarNames = this._getSimilarContractNames(contractName, names);
+
+    if (similarNames.length === 1) {
+      throw new HardhatError(ERRORS.ARTIFACTS.TYPO_SUGGESTION, {
```

Comment:
@alcuadrado what do you think about this? I think the options are:

- Add a new error code for the scenario where only one suggestion is available, with the message "Did you mean..." instead of the list of suggestions.
- A single error with a flexible "suffix" message that has three possible scenarios.

The second option adds some complexity and makes the `throw` sites uglier. But OTOH the underlying error is the same, so you could argue that having different codes for each one just for the message might be a bad idea long therm.

I think Morgan prefers using a different error for each case. I'm on the fence, but have a weak preference for using a single error.

Reply:
No strong opinion, but i think i'd go with the second approach. The reason is that I think using the same error code valuable. For example, it will result in more hits when you google it.

Reply:
My thoughts are that I don't think we necessarily need to have different messaging for a single suggestion vs. multiple. From my (admittedly green) point of view, a big benefit to how our errors are structured is that you can throw a specific error with the parameters it needs and not have to deal with any of the output formatting because all of that is abstracted away. 

If we try to force this error into the existing one (rather than making a new error code) by just tacking on a catch-all variable to the end, I think that introduces a slippery-slope precedent for future cases like this, and defeats the purpose of the current error design.

That being said, I think it's worth mentioning that the current artifact not found error is probably not needed in its current form anymore. The only time its being used now is if the artifact directory is empty when they try to read artifacts from it, so a case could be made that artifact not found would be better suited for this suggestion feature and a new error could be used for the singular use case of an empty artifact directory maybe? I'm not sure.

Reply:
> If we try to force this error into the existing one (rather than making a new error code) by just tacking on a catch-all variable to the end, I think that introduces a slippery-slope precedent for future cases like this, and defeats the purpose of the current error design.

I agree with this, but I also think that the proper solution would be to (eventually) make the message-generation code smarter instead of a search-and-replace thing. I wouldn't do that in this PR though, but next time we run into something like we can refactor the error list to be a bit more powerful.

Another reason I prefer the single-code approach is that if we don't do it and later we change our minds, we'll need to use a single error and deprecate the other two.

Again, weak preference! If these arguments aren't compelling enough for you, I'm ok with just adding new errors. But I do think we need the "Did you mean..." thing when there's only one suggestion.

Reply:
That's fair. I can get on board with editing the existing code for now and just kicking the can down the road until we refactor the error system.